aspnetcore/security/authentication/windowsauth/6.0samples/WebRPwinAuth/Program.cs

```diff
@@ -0,0 +1,225 @@
+                    settings.MachineAccountName = "machineName";
+                    settings.MachineAccountPassword =
+                                      builder.Configuration["Password"];
+                });
+            }
+        });
+
+builder.Services.AddRazorPages();
+#endregion
+
+var app = builder.Build();
+
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+app.UseAuthentication();
+app.UseAuthorization();
+
+app.MapRazorPages();
+
+app.Run();
+#elif HTS
+#region snippet_hts
+using Microsoft.AspNetCore.Server.HttpSys;
+using System.Runtime.InteropServices;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddAuthentication(HttpSysDefaults.AuthenticationScheme);
+
+if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+{
+    builder.WebHost.UseHttpSys(options =>
+        {
+            options.Authentication.Schemes =
+                AuthenticationSchemes.NTLM |
+                AuthenticationSchemes.Negotiate;
+            options.Authentication.AllowAnonymous = false;
+        });
+}
+#endregion
+
+builder.Services.AddRazorPages();
+
+var app = builder.Build();
+
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+app.UseAuthentication();
+app.UseAuthorization();
+
+app.MapRazorPages();
+
+app.Run();
+#elif IMP
+#pragma warning disable CA1416 // Validate platform compatibility
+using Microsoft.AspNetCore.Authentication.Negotiate;
+using System.Security.Principal;
+using System.Text;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddAuthentication(NegotiateDefaults.AuthenticationScheme)
+   .AddNegotiate();
+
+builder.Services.AddAuthorization(options =>
+{
+    options.FallbackPolicy = options.DefaultPolicy;
+});
+builder.Services.AddRazorPages();
+
+var app = builder.Build();
+
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+app.UseAuthentication();
+app.UseAuthorization();
+
+app.MapRazorPages();
+#region snippet_imp
+app.Run(async (context) =>
+{
+    try
+    {
+        var user = (WindowsIdentity)context.User.Identity!;
+
+        await context.Response
+            .WriteAsync($"User: {user.Name}\tState: {user.ImpersonationLevel}\n");
+
+        WindowsIdentity.RunImpersonated(user.AccessToken, () =>
+        {
+            var impersonatedUser = WindowsIdentity.GetCurrent();
+            var message =
+                $"User: {impersonatedUser.Name}\t" +
+                $"State: {impersonatedUser.ImpersonationLevel}";
+
+            var bytes = Encoding.UTF8.GetBytes(message);
+            context.Response.Body.Write(bytes, 0, bytes.Length);
+        });
```

Comment:
There's a new async API for this:
```suggestion
        await WindowsIdentity.RunImpersonatedAsync(user.AccessToken, async () =>
        {
            var impersonatedUser = WindowsIdentity.GetCurrent();
            var message =
                $"User: {impersonatedUser.Name}\t" +
                $"State: {impersonatedUser.ImpersonationLevel}";

            var bytes = Encoding.UTF8.GetBytes(message);
            await context.Response.Body.WriteAsync(bytes, 0, bytes.Length);
        });
```